nakadi-java-client/src/test/java/nakadi/ExponentialRetryTest.java

```diff
@@ -8,81 +8,88 @@
         .percentOfMaxIntervalForJitter(20)
         .maxTime(3, TimeUnit.SECONDS)
         .build();
+    runRetries(exponentialRetry);
+    validateTimeoutState(exponentialRetry);
+  }
 
-    while(true) {
-      long l = exponentialRetry.nextBackoffMillis();
-      if(l == -1) {
-        break;
-      }
-      Thread.sleep(l);
-    }
-
-    assertTrue(exponentialRetry.workingTime >= exponentialRetry.maxTime);
-    assertTrue(exponentialRetry.workingAttempts < exponentialRetry.maxAttempts);
-
-    exponentialRetry = ExponentialRetry.newBuilder()
+  @Test
+  public void tempusFugit_2() throws Exception {
+    ExponentialRetry exponentialRetry = ExponentialRetry.newBuilder()
         .maxTime(3, TimeUnit.SECONDS)
         .maxInterval(100, TimeUnit.MILLISECONDS)
         .build();
+    runRetries(exponentialRetry);
+    validateTimeoutState(exponentialRetry);
+  }
 
-    while(true) {
-      long l = exponentialRetry.nextBackoffMillis();
-      if(l == -1) {
-        break;
-      }
-      Thread.sleep(l);
-    }
-
-    assertTrue(exponentialRetry.workingTime >= exponentialRetry.maxTime);
+  private void validateTimeoutState(ExponentialRetry exponentialRetry) {
+    assertTrue(exponentialRetry.workingTime() >= exponentialRetry.maxTime);
     assertTrue(exponentialRetry.workingAttempts < exponentialRetry.maxAttempts);
   }
 
   @Test
-  public void annumero() throws Exception {
-
+  public void annumero_1() throws Exception {
     ExponentialRetry exponentialRetry = ExponentialRetry.newBuilder()
         .initialInterval(101, TimeUnit.MILLISECONDS)
         .maxAttempts(20)
         .maxInterval(100, TimeUnit.MILLISECONDS)
         .maxTime(Integer.MAX_VALUE, TimeUnit.SECONDS)
         .build();
+    runRetries(exponentialRetry);
+    validateRetriesExceededState(exponentialRetry);
+  }
 
-    while(true) {
-      long l = exponentialRetry.nextBackoffMillis();
-      if(l == -1) {
-        break;
-      }
-      Thread.sleep(l);
-    }
-
-    assertTrue(exponentialRetry.workingTime < exponentialRetry.maxTime);
-    assertTrue(exponentialRetry.workingAttempts >= exponentialRetry.maxAttempts);
+  @Test
+  public void annumero_2() throws Exception {
+    ExponentialRetry exponentialRetry = ExponentialRetry.newBuilder()
+            .maxAttempts(3)
+            .maxInterval(100, TimeUnit.MILLISECONDS)
+            .build();
+    runRetries(exponentialRetry);
+    validateRetriesExceededState(exponentialRetry);
+  }
 
+  @Test
+  public void workingTimeCalculation() {
+    ExponentialRetry exponentialRetry = ExponentialRetry.newBuilder()
+            .maxInterval(100, TimeUnit.MILLISECONDS)
+            .maxTime(110, TimeUnit.MILLISECONDS)
+            .build();
+    exponentialRetry.nextBackOffMillis(1);
+    assertFalse(exponentialRetry.isFinished());
+    exponentialRetry.nextBackOffMillis(100);
+    assertFalse(exponentialRetry.isFinished());
+    exponentialRetry.nextBackOffMillis(101);
+    assertFalse(exponentialRetry.isFinished());
```

Comment:
Would previously fail here as accumulated time was 202 milliseconds (instead of 101).